### PR TITLE
AirfRANS: volume smoothness regularization to cut vol_mse

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Smooth volume field predictions should reduce `vol_mse`. The current champion (PR #3135) achieves `vol_mse=0.002039` but physical intuition tells us CFD volume fields are spatially smooth — adjacent points in the domain share similar pressure/velocity values. By penalizing spatial discontinuities in predicted volume fields, we may teach the model to respect this physical prior and reduce `vol_mse` toward <0.0017.

Proposed smoothness regularization: for each batch of volume points, find k-nearest neighbours among the volume nodes (k=4 or k=8), then penalize large prediction differences between adjacent pairs.

## Instructions

Start from the champion config (PR #3135):

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0
```

**Step 0: smoke test first.** Run 3 epochs on Trial 1 (below) before committing to full runs. Confirm training doesn't diverge and vol smoothness loss is non-zero.

### Implementation — Volume Smoothness Regularizer

Add the following logic to `train.py` (in the volume-node loss computation block):

1. **Find neighbours** (do this on the batch of volume node positions `pos_vol`):

```python
from torch_geometric.nn import knn_graph  # or use torch cdist + topk

# k=4 neighbours within the volume batch
edge_index = knn_graph(pos_vol, k=4, batch=batch_vol, loop=False)
src, dst = edge_index[0], edge_index[1]
```

2. **Compute smoothness loss** — two formulations to try:

**Formulation A — difference-of-differences (physics-aware):**
```python
# Penalise predictions that are *more* different than the target is
smooth_loss = ((pred_vol[src] - pred_vol[dst])**2 - (target_vol[src] - target_vol[dst])**2).pow(2).mean()
```

**Formulation B — strict smoothness:**
```python
# Penalise any prediction difference between neighbours
smooth_loss = ((pred_vol[src] - pred_vol[dst])**2).mean()
```

3. **Add to total loss:**
```python
total_loss = primary_loss + lambda_smooth * smooth_loss
```

### Trial configurations

Add a `--vol-smooth-lambda` argument (default 0.0) and a `--vol-smooth-k` argument (default 4). Use `--wandb_group af-volume-smooth-regularization` on every run.

| Trial | Formulation | lambda | k |
|-------|-------------|--------|---|
| 1     | A (diff-of-diff) | 0.01 | 4 |
| 2     | A (diff-of-diff) | 0.001 | 4 |
| 3     | B (strict) | 0.001 | 4 |
| 4     | B (strict) | 0.0001 | 8 |

Run all four trials to full convergence (or until the `SENPAI_TIMEOUT_MINUTES` limit).

**Hard constraints:**
- `surface_mse` must NOT regress beyond 0.000296 (the current baseline). If any trial increases surface_mse, discard it regardless of vol_mse.
- Target: `vol_mse < 0.002039`, ideally `vol_mse < 0.0017`.

Report the best `surface_mse` and `vol_mse` (from best val checkpoint) for each trial, plus W&B run IDs.

## Baseline

Current AirfRANS champion metrics (PR #3135, nezuko):
- `val_primary/surface_mse`: **0.000296** (epoch 704)
- `val_primary/vol_mse`: **0.002039** (epoch 743)
- Baseline W&B run: `sh2zyfwr`
- Config: 2L/256d/4H, AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, vol-weight=10x, Fourier

Reproduce command:
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT norman:

## Results

All 4 trials **regressed** on both `surface_mse` and `vol_mse` relative to the baseline. The volume smoothness regularization did not help.

### Summary Table

| Trial | Formulation | λ | k | Val surface_mse | Val vol_mse | Test surface_mse | Test vol_mse | Best Epoch | W&B Run |
|-------|-------------|------|---|-----------------|-------------|------------------|--------------|------------|---------|
| **Baseline** | — | — | — | **0.000296** | **0.002039** | — | — | 704/743 | `sh2zyfwr` |
| 1 | A (diff-of-diff) | 0.01 | 4 | 0.000598 (+102%) | 0.005337 (+162%) | 0.001881 | 0.006112 | 344 | `6oo3dn4f` |
| 2 | A (diff-of-diff) | 0.001 | 4 | 0.000490 (+66%) | 0.004066 (+99%) | 0.000854 | 0.004856 | 396 | `q2zyfwfg` |
| 3 | B (strict) | 0.001 | 4 | 0.000491 (+66%) | 0.004949 (+143%) | 0.000885 | 0.006040 | 345 | `6dqv5rod` |
| 4 | B (strict) | 0.0001 | 8 | 0.000575 (+94%) | 0.004996 (+145%) | 0.000959 | 0.006002 | 340 | `wr3xsn6u` |

**Hard constraint check:** ALL trials violate the surface_mse regression threshold (0.000296). None are viable.

### Commands Used

```bash
# Trial 1
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 --vol-smooth-lambda 0.01 --vol-smooth-k 4 --vol-smooth-formulation A --agent norman --wandb-name "norman/af-vol-smooth-t1-formA-lam0.01-k4" --wandb-group "af-volume-smooth-regularization"

# Trial 2
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 --vol-smooth-lambda 0.001 --vol-smooth-k 4 --vol-smooth-formulation A --agent norman --wandb-name "norman/af-vol-smooth-t2-formA-lam0.001-k4" --wandb-group "af-volume-smooth-regularization"

# Trial 3
CUDA_VISIBLE_DEVICES=2 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 --vol-smooth-lambda 0.001 --vol-smooth-k 4 --vol-smooth-formulation B --agent norman --wandb-name "norman/af-vol-smooth-t3-formB-lam0.001-k4" --wandb-group "af-volume-smooth-regularization"

# Trial 4
CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 --vol-smooth-lambda 0.0001 --vol-smooth-k 8 --vol-smooth-formulation B --agent norman --wandb-name "norman/af-vol-smooth-t4-formB-lam0.0001-k8" --wandb-group "af-volume-smooth-regularization"
```

### What Happened

**The hypothesis failed.** Both formulations (A: diff-of-differences, B: strict smoothness) degraded performance across all lambda/k settings tested.

Key observations:
- **Surface prediction hurt most**: Even the weakest regularization (λ=0.0001) increased surface_mse by ~94%, suggesting the smoothness penalty introduces gradient interference that destabilizes surface learning.
- **Volume prediction also regressed**: vol_mse roughly doubled across all trials, the opposite of the intended effect.
- **Higher lambda = worse**: Trial 1 (λ=0.01) was worst, confirming the regularizer is actively harmful — not just too weak.
- **Both formulations equivalent**: Formulation A (physics-aware diff-of-diff) and B (strict smoothness) performed similarly poorly, suggesting the issue is with KNN-based smoothness as a concept, not the specific formulation.
- **Early convergence**: All trials peaked around epoch 340-396, whereas the baseline peaked at 704-743, suggesting the regularizer disrupts late-stage fine-tuning.

Why it likely failed:
1. **CFD volume fields aren't uniformly smooth** — boundary layers, shear regions, and wake structures have sharp physical gradients that the smoothness penalty incorrectly penalizes.
2. **Random subsampling** (capped at 4000 nodes) + brute-force `cdist` may not capture physically meaningful neighborhoods — true neighbors in the CFD mesh topology aren't necessarily the nearest in Euclidean space.
3. **Gradient competition**: The smoothness loss competes with the primary volume MSE objective, creating a Pareto conflict that degrades both metrics.

### Suggested Follow-ups

- **Mesh-topology-aware regularization**: Instead of KNN in Euclidean space, use the actual mesh connectivity (edges from the CFD mesh) to define neighborhoods. This would respect the physics better.
- **Gradient penalty on volume fields**: Instead of penalizing prediction differences, penalize the spatial gradient magnitude of the *residual* (pred - target), which would only smooth errors, not legitimate sharp features.
- **Anisotropic smoothness**: Apply smoothness only in directions perpendicular to the flow (cross-stream), not along sharp gradients.
- **Volume loss weighting schedule**: Rather than adding a new loss term, explore scheduling the existing `volume_loss_weight` to increase later in training.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3223 because ADVISOR: Closing — volume smoothness regularization is a dead end for AF. All 4 trials regressed BOTH metrics: surface 44-94% worse, volume 60-97% worse. Even the smallest lambda (1e-4) is harmful. Root cause: KNN Euclidean smoothness is physically misaligned with AF flow fields — boundary layers, wakes, and shear layers have legitimate sharp spatial gradients. The penalty suppresses real physics. Additionally, gradient interference from the shared backbone means the volume regularizer corrupts surface gradients too. Your diagnosis was spot-on — mesh-topology-aware or residual-gradient approaches would avoid these problems but are different experiments..
